scene/resources/fog_material.cpp

```diff
@@ -0,0 +1,205 @@
+	return height_offset;
+}
+
+void FogMaterial::set_height_falloff(float p_falloff) {
+	height_falloff = p_falloff;
+	RS::get_singleton()->material_set_param(_get_material(), "height_falloff", height_falloff);
+}
+
+float FogMaterial::get_height_falloff() const {
+	return height_falloff;
+}
+
+void FogMaterial::set_edge_fade(float p_edge_fade) {
+	edge_fade = p_edge_fade;
+	RS::get_singleton()->material_set_param(_get_material(), "edge_fade", edge_fade);
+}
+
+float FogMaterial::get_edge_fade() const {
+	return edge_fade;
+}
+
+void FogMaterial::set_density_texture(const Ref<Texture3D> &p_texture) {
+	density_texture = p_texture;
+	RID tex_rid = p_texture.is_valid() ? p_texture->get_rid() : RID();
+	RS::get_singleton()->material_set_param(_get_material(), "density_texture", tex_rid);
+}
+
+Ref<Texture3D> FogMaterial::get_density_texture() const {
+	return density_texture;
+}
+
+Shader::Mode FogMaterial::get_shader_mode() const {
+	return Shader::MODE_FOG;
+}
+
+RID FogMaterial::get_shader_rid() const {
+	_update_shader();
+	return shader;
+}
+
+RID FogMaterial::get_rid() const {
+	_update_shader();
+	if (!shader_set) {
+		RS::get_singleton()->material_set_shader(_get_material(), shader);
+		shader_set = true;
+	}
+	return _get_material();
+}
+
+bool FogMaterial::_can_do_next_pass() const {
+	return false;
+}
+
+void FogMaterial::_validate_property(PropertyInfo &property) const {
+	if (property.name == "render_priority" || property.name == "next_pass") {
+		property.usage = PROPERTY_USAGE_NONE;
+	}
+}
+
+void FogMaterial::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_density", "density"), &FogMaterial::set_density);
+	ClassDB::bind_method(D_METHOD("get_density"), &FogMaterial::get_density);
+	ClassDB::bind_method(D_METHOD("set_albedo", "albedo"), &FogMaterial::set_albedo);
+	ClassDB::bind_method(D_METHOD("get_albedo"), &FogMaterial::get_albedo);
+	ClassDB::bind_method(D_METHOD("set_emission", "emission"), &FogMaterial::set_emission);
+	ClassDB::bind_method(D_METHOD("get_emission"), &FogMaterial::get_emission);
+	ClassDB::bind_method(D_METHOD("set_height_offset", "height_offset"), &FogMaterial::set_height_offset);
+	ClassDB::bind_method(D_METHOD("get_height_offset"), &FogMaterial::get_height_offset);
+	ClassDB::bind_method(D_METHOD("set_height_falloff", "height_falloff"), &FogMaterial::set_height_falloff);
+	ClassDB::bind_method(D_METHOD("get_height_falloff"), &FogMaterial::get_height_falloff);
+	ClassDB::bind_method(D_METHOD("set_edge_fade", "edge_fade"), &FogMaterial::set_edge_fade);
+	ClassDB::bind_method(D_METHOD("get_edge_fade"), &FogMaterial::get_edge_fade);
+	ClassDB::bind_method(D_METHOD("set_density_texture", "density_texture"), &FogMaterial::set_density_texture);
+	ClassDB::bind_method(D_METHOD("get_density_texture"), &FogMaterial::get_density_texture);
+
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "density", PROPERTY_HINT_RANGE, "0.0,16.0,0.01,or_greator,or_lesser"), "set_density", "get_density");
```

Comment:
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "density", PROPERTY_HINT_RANGE, "0.0,16.0,0.0001,or_greater,or_lesser"), "set_density", "get_density");
```
to match with https://github.com/godotengine/godot/blob/1f1f8b34d705af1cda9d38d6aa6550134aee860f/scene/resources/environment.cpp#L1348. Helps with making subtler large volumes.
Personally I wouldn't have a problem with an additional zero beyond for both of these properties, but I'm making exceptionally large scenes.